ddspipe_core/src/cpp/communication/dds/DdsBridge.cpp

```diff
@@ -175,6 +97,178 @@ void DdsBridge::disable() noexcept
+            // The writer already exists. There is nothing to do. Exit.
+            return;
+        }
+    }
+
+    std::shared_ptr<IParticipant> participant = participants_->get_participant(discoverer_participant_id);
+
+    // Create the writer.
+    std::map<ParticipantId, std::shared_ptr<IWriter>> writer;
+    writer[discoverer_participant_id] = participant->create_writer(*topic_);
+
+    // Add the writer to the tracks it has routes for.
+    add_writers_to_tracks_(writer);
+}
+
+void DdsBridge::remove_from_tracks(
+        const ParticipantId& discoverer_participant_id) noexcept
+{
+    for (auto it = tracks_.cbegin(), next_it = it; it != tracks_.cend(); it = next_it)
+    {
+        ++next_it;
+
+        const auto& track = it->second;
+
+        // If the writer is in the track, remove it.
+        track->remove_writer(discoverer_participant_id);
+
+        if (track->count_writers() <= 0)
+        {
+            // The track doesn't have any writers. Remove it.
+            tracks_.erase(it);
+        }
+    }
+}
+
+void DdsBridge::add_writers_to_tracks_(
+        std::map<ParticipantId, std::shared_ptr<IWriter>>& writers)
+{
+    // Add writers to the tracks of the readers in their route.
+    // If the readers in their route don't exist, create them with their tracks.
+    for (const ParticipantId& id : participants_->get_participants_ids())
+    {
+
+        // Select the necessary writers
+        std::map<ParticipantId, std::shared_ptr<IWriter>> writers_of_reader;
+
+        const auto& routes_it = routes_.find(id);
+
+        if (routes_it != routes_.end())
+        {
+            // The reader has a route. Add only the writers in the route.
+            const auto& writers_in_route_of_reader = routes_it->second;
+
+            for (const auto& writer_id : writers_in_route_of_reader)
+            {
+                writers_of_reader[writer_id] = writers[writer_id];
+            }
+        }
+        else
+        {
+            // The reader doesn't have a route. Add every writer (+ itself if repeater)
+            writers_of_reader = writers;
+
+            if (!participants_->get_participant(id)->is_repeater())
+            {
+                // The participant is not a repeater. Do not add its writer.
+                writers_of_reader.erase(id);
+            }
+        }
+
+        if (writers_of_reader.size() == 0)
+        {
+            // There are no writers in the route. Don't create the reader or the track.
+            continue;
+        }
+
+        if (tracks_.count(id))
+        {
+            // The track already exists. Add the writers to it.
+            for (const auto& writers_of_reader_it : writers_of_reader)
+            {
+                const auto& writer_id = writers_of_reader_it.first;
+                const auto& writer = writers_of_reader_it.second;
+
+                if (!tracks_[id]->has_writer(writer_id))
+                {
+                    // Enable the writer before adding it to the track.
```

Comment:
Be careful here, enabling/disabling writers and readers should be done by track depending on its status.